feat: 查询支持下载接口 --story=125200073

Zhaoyikaiii opened the pull request.

Zhaoyikaiii committed:
feat: add support for scroll queries

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,387 @@
+	Query       *metadata.Query `json:"query"`
+	Type        string          `json:"type"`
+	Offset      int64           `json:"offset"`
+	ScrollIDs   []string        `json:"scroll_ids"`
```
这个字段应该合并到 SliceStates 里面，同时兼容多个 es 的情况

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,387 @@
+		"username": username,
+	}
+
+	queryBytes, err := json.Marshal(keyMap)
```
这里需要看下是否有顺序

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,387 @@
+
+import (
+	"context"
+	"encoding/json"
```
用另一个 json 库

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,387 @@
+	result := client.Get(ctx, sessionKey)
+	if result.Err() == nil {
+		var session SessionObject
+		if err := json.Unmarshal([]byte(result.Val()), &session); err == nil {
```
获取到了之后需要续期

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -328,8 +341,22 @@ func (i *Instance) esQuery(ctx context.Context, qo *queryOption, fact *FormatFac
+			// 检查是否有 slice 信息传递
+			sliceID, maxSlice, hasSlice := i.parseSliceInfo(qb.ScrollIDs)
+
+			if hasSlice {
```
这里的写法优雅一点：

ss := client.Scroll(qo.indexes...).Scroll(qb.Scroll).SearchSource(source)
			if true {
				ss.Slice(elastic.NewSliceQuery().Id(0).Max(3))
			}

chenjiandongx commented on pkg/unify-query/internal/set/set.go:
```diff
@@ -31,6 +31,9 @@ func New[T comparable](items ...T) *Set[T] {
 func (s *Set[T]) First() (v T) {
 	s.lock.RLock()
 	defer s.lock.RUnlock()
+	if s.m == nil {
```
会为什么会出现不使用 New 而直接使用的场景？该场景是否符合预期？

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,189 @@
+
+// GetNextScrollID 获取下一个可用的scrollID(ES)或返回当前index(Doris)
+func (s *ScrollSession) GetNextScrollID(ctx context.Context, tsDbType, connect, tableID string, sliceIdx int) (string, int, error) {
+	if tsDbType == "elasticsearch" {
```
编写代码尽量不要在一个 if 里实现太多逻辑，遵循类似错误优先原则。

优先处理逻辑较少的分支。而且为什么 ES 是 tsdb?

Zhaoyikaiii replied:
已修改

tencentblueking-adm commented:
[![CLA assistant check](https://bk-cla.bktencent.com:443/pull/badge/signed)](https://bk-cla.bktencent.com:443/TencentBlueKing/bkmonitor-datalink?pullRequest=853) <br/>All committers have signed the CLA.

shamcleren commented on pkg/unify-query/metadata/result_table_options.go:
```diff
@@ -15,6 +15,9 @@ type ResultTableOption struct {
 	SearchAfter []any  `json:"search_after,omitempty"`
+	SliceID     *int   `json:"slice_id,omitempty"`
+	SliceMax    *int   `json:"slice_max,omitempty"`
+	Connect     string `json:"connect,omitempty"`
```
这里要看下用途，是否可以去掉

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/metadata/result_table_options.go:
```diff
@@ -15,6 +15,8 @@ type ResultTableOption struct {
 	ScrollID    string `json:"scroll_id,omitempty"`
 	SearchAfter []any  `json:"search_after,omitempty"`
+	SliceID     *int   `json:"slice_id,omitempty"`
+	SliceMax    *int   `json:"slice_max,omitempty"`
```
SliceMax 这个好像不是用户配置的？而且 Slice 的特性，我们直接闭环了，不需要用户介入吧？

Zhaoyikaiii replied:
这里是因为会有循环依赖.会先在外面通过option传进去.然后到了storage再用option里面的值

chenjiandongx replied:
那这几个 option 需要被 json 序列化吗？

Zhaoyikaiii replied:
需要,有一些测试用例的序列化需要依赖这里的json tag

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,166 @@
+)
+
+func (s *ScrollSession) GetNextScrollID(ctx context.Context, storageType, connect, tableID string, sliceIdx int) (string, int, error) {
+	if storageType != consul.ElasticsearchStorageType {
```
这里不是应该判定是不是 doris 吗？直接用 switch case 好一点，如果不在这几种情况，应该要报异常吧？

Zhaoyikaiii replied:
已修改:https://github.com/TencentBlueKing/bkmonitor-datalink/pull/853/commits/555f079034681321182e5b214adfe2e9a96b1653

Zhaoyikaiii committed:
refactor: simplify ScrollSession structure and update related logic for scrollID management

Zhaoyikaiii committed:
chore:如果已经done,继续请求不返回报错

Zhaoyikaiii committed:
chore: simplify result option

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,300 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+						defer sendWg.Done()
+
+						sessionLock.Lock()
+						scrollID, index, err := session.GetNextScrollID(ctx, instance.InstanceType(), connect, tableId, currentSliceIndex)
```
GetNextScrollID 这个命名有歧义，应该是 CurrentScrollID 吧？或者直接写 GetScrollID 也可以

Zhaoyikaiii replied:
已修改:[chore: change function naming: GetNextScrollID -> CurrentScrollID](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/853/commits/f12332e32a8ad13298f9e366e97b56197912a98e)

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,300 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+							lock.Unlock()
+						}
+
+						if len(queryList) > 1 {
```
这一段是查询原始数据翻页的处理逻辑，在下载场景应该不需要了？

Zhaoyikaiii replied:
已修改:@Zhaoyikaiii
[refactor: remove redundant size adjustment for queryList in queryRawW](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/853/commits/b926cb1a4ac78eb6e147416a2049fcbf6be454ad)

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,300 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+							SliceMax: &session.MaxSlice,
+							ScrollID: scrollID,
+						}
+						sliceQuery.From = index * session.Limit
```
这里的赋值是直接赋值了？是不是需要考虑失败的场景？

Zhaoyikaiii replied:
已修改:@Zhaoyikaiii
[fix: handle invalid index in queryRawWithScroll function to prevent n…](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/853/commits/9f1db8142e415f8c3bb8f776263c2ccf023ea256)

Zhaoyikaiii committed:
feat: add support for handling unsupported storage types in ScrollSession

Zhaoyikaiii committed:
chore: change function naming: GetNextScrollID -> CurrentScrollID

Zhaoyikaiii committed:
refactor: remove redundant size adjustment for queryList in queryRawWithScroll function

Zhaoyikaiii committed:
fix: handle invalid index in queryRawWithScroll function to prevent negative slice queries

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,138 @@
+	SessionStatusFailed  = "FAILED"
+)
+
+func (s *ScrollSession) CurrentScrollID(ctx context.Context, storageType, connect, tableID string, sliceIdx int) (string, int, error) {
```
提供一个 New 方法并在里面进行初始化，不需要每个方法下面都进行初始化。

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,138 @@
+			return true
+		}
+
+		connectTablePairs := make(map[string]bool)
```
使用 map 进行去重的时候，key 不一定是 string，它也可以是 struct。另外，如果 map 的 values 是没有意义的话，官方的建议是使用 struct{} 来替代。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/redis/scroll_integration_test.go:
```diff
@@ -0,0 +1,566 @@
+
+	log.Infof(ctx, "[TEST] Session ScrollIDs after round 1: %+v", session.ScrollIDs)
+
+	// 模拟第二轮查询：每个slice又获取到新的scrollID（这里可能是问题所在）
```
但是你的第一个测试也是使用的相同的 scrollid，为什么到这里就变成了问题所在？

chenjiandongx commented on pkg/unify-query/redis/scroll_integration_test.go:
```diff
@@ -0,0 +1,566 @@
+	log.Infof(ctx, "[TEST] ====== 问题流程测试完成 - 现在应该正确处理 ======")
+}
+
+// TestESScrollFixedFlow 测试修复后的ES scroll流程
```
==，这里是 AI 修复你的上一个出问题的单测，然后又搞了个新的？或者说这个和第一个测试有什么区别？

chenjiandongx commented on pkg/unify-query/redis/scroll_integration_test.go:
```diff
@@ -0,0 +1,566 @@
+	assert.False(t, session.HasMoreData("elasticsearch"), "Should not have more data")
+	assert.Equal(t, SessionStatusDone, session.Status, "Session should be done")
+
+	log.Infof(ctx, "[TEST] ====== 测试完成 ======")
```
没必要在每个测试后面都输出一个测试完成，没有出错自然就完成了。

chenjiandongx commented on pkg/unify-query/redis/scroll_integration_test.go:
```diff
@@ -0,0 +1,566 @@
+}
+
+// TestESScrollCorrectTermination 测试正确的ES scroll终止逻辑
+func TestESScrollCorrectTermination(t *testing.T) {
```
这个感觉也跟第一个测试是一个东西。

chenjiandongx commented on pkg/unify-query/redis/scroll_integration_test.go:
```diff
@@ -0,0 +1,566 @@
+		err = ScrollProcessSliceResult(ctx, &session, connect, tableID, "", sliceIdx, "elasticsearch", 5, options)
+		assert.NoError(t, err)
+
+		log.Infof(ctx, "[TEST] Slice %d round 1: added scrollID %s", sliceIdx, newScrollID)
```
另外，所有的测试 log 建议使用 `t.Log` 或者 `t.Logf`，同时也没有必要在每个 log 语句之前都写一个 TEST。

chenjiandongx commented on pkg/unify-query/redis/scroll_key_test.go:
```diff
@@ -0,0 +1,172 @@
+			},
+		},
+		{
+			name:     "different username",
```
这个单测和第一个单测有什么区别吗？

chenjiandongx commented on pkg/unify-query/redis/scroll_key_test.go:
```diff
@@ -0,0 +1,172 @@
+
+			// 验证包含正确的字段
+			assert.Contains(t, keyMap, "queryTs")
+			assert.Contains(t, keyMap, "username")
```
这个断言看起来没意义，你在下一个断言都已经要求【存在且相等了】。

chenjiandongx commented on pkg/unify-query/redis/scroll_key_test.go:
```diff
@@ -0,0 +1,172 @@
+			assert.Equal(t, tc.username, keyMap["username"])
+
+			// 验证queryTs字段不为空
+			assert.NotNil(t, keyMap["queryTs"])
```
这个同理。

chenjiandongx commented on pkg/unify-query/redis/scroll_key_test.go:
```diff
@@ -0,0 +1,172 @@
+	key2, err := ScrollGenerateQueryTsKey(queryTs, "user2")
+	assert.NoError(t, err)
+
+	assert.NotEqual(t, key1, key2, "不同用户应该生成不同的键")
```
建议统一 assert 的 desc 风格。

chenjiandongx commented on pkg/unify-query/redis/scroll_key_test.go:
```diff
@@ -0,0 +1,172 @@
+	assert.Equal(t, key1, key3, "相同用户和查询应该生成相同的键")
+}
+
+func TestScrollGenerateQueryTsKeyWithDifferentQueryTs(t *testing.T) {
```
感觉这个单测可以和上一个单测合并。

chenjiandongx commented on pkg/unify-query/redis/scroll_key_test.go:
```diff
@@ -0,0 +1,172 @@
+	// 测试会话键生成
+	sessionKey := GetSessionKey(queryTsKey)
+	assert.True(t, len(sessionKey) > len(SessionKeyPrefix))
+	assert.Contains(t, sessionKey, SessionKeyPrefix)
```
同理，你的第二个断言已经覆盖了第一个断言。

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+	"github.com/TencentBlueKing/bkmonitor-datalink/pkg/unify-query/metadata"
+)
+
+var ScrollGenerateQueryTsKey = func(queryTs any, userName string) (string, error) {
```
是出于什么考虑需要把函数的定义写成 `var FooBar = func()` 而不是正常的 `func FooBar()`？

Zhaoyikaiii replied:
这里是看到这个包其他的函数也是这样写的,为了保持一样的风格

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+	return string(queryBytes), nil
+}
+
+var GetSessionKey = func(queryTsKey string) string {
```
同上。

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+	queryBytes, err := json.Marshal(keyStruct)
+	if err != nil {
+		log.Errorf(
+			context.Background(),
```
任何返回 err 的地方就不要再一次 log，这样上层拿到后大概率又会再打一次 log。

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+	}
+	if forceClear {
+		if err := client.Del(ctx, sessionKey).Err(); err != nil {
+			log.Warnf(ctx, "failed to clear session: %v", err)
```
如果删除失败，继续执行下面的流程是符合预期的吗？

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+			return session, nil
+		}
+	}
+
```
如果这里的 reuslt 返回是 Err，也是需要义无反顾地继续往下执行吗？

Zhaoyikaiii replied:
- 如果result.Err()是nil代表已经存在,更新session就可以
- 如果是非redisNilErr则是其他的redis的报错直接返回
- 如果是redisNilErr则是没有这个session在后续会创建这个session

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+
+	err = client.Set(ctx, sessionKey, sessionBytes, session.ScrollTimeout).Err()
+	if err != nil {
+		log.Errorf(ctx, "Failed to update session %s: %v", sessionKey, err)
```
同上。

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+var ScrollProcessSliceResult = func(ctx context.Context, session *ScrollSession, connect, tableID, scrollID string, sliceIdx int, tsDbType string, size int64, options metadata.ResultTableOptions) error {
+	isDone := size == 0
+
+	if tsDbType == consul.ElasticsearchStorageType {
```
是不是优先判断 isDone 会好一些。其次，isDone 是一个确定的结果，为什么代码是优先判断 newScrollID 是否为空再来判断 `else if isDone`，这不是一个合理的选择。

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,200 @@
+			log.Debugf(ctx, "slice %d completed for %s|%s (empty result and no new scrollID)", sliceIdx, connect, tableID)
+		}
+
+		hasMoreData := session.HasMoreData("elasticsearch")
```
另外，整个 PR 写了 10+ 处 `session.HasMoreData("elasticsearch")`，这是意味着 `elasticsearch` 本身是一个常量？

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+
+	list = make([]map[string]any, 0)
+
+	if queryTs.SpaceUid == "" {
```
建议跟上面的校验逻辑放在一起。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+	sessionKey := redisUtil.GetSessionKey(queryTsKey)
+	lockKey := redisUtil.GetLockKey(queryTsKey)
+	scrollWindowDuration, _ := time.ParseDuration(queryTs.Scroll)
+	lockTimeoutDuration := ScrollLockTimeout
```
这个赋值可以去掉，你用了一个更长的名称来替换一个本来也不短的名称。

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+
+	queryTsKey, err := redisUtil.ScrollGenerateQueryTsKey(queryTs, user.Name)
+	if err != nil {
+		return 0, nil, nil, false, fmt.Errorf("failed to generate queryTs key: %v", err)
```
这个 err 到底是不是函数定义时的 err，不要把两种 err 返回值混用。

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+		}
+
+		for _, storageId := range storageIds {
+			if storageQueryMap[storageId] == nil {
```
Tip: 这里不初始化也不会有问题。

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
 	return resp, nil
 }
+
+func queryRawWithScroll(ctx context.Context, queryTs *structured.QueryTs) (total int64, list []map[string]any, resultTableOptions metadata.ResultTableOptions, done bool, err error) {
```
这个函数的复杂度太高了，建议抽象一下。代码都倾斜成 45° 了，几乎没有什么可读性。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+
+						if index >= 0 {
+							sliceQuery.From = index * session.Limit
+						} else {
```
所以会出现 index 小于 0 的情况？

Zhaoyikaiii replied:
不会,已移除

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+						sliceQuery.ResultTableOptions.SetOption(sliceQuery.TableID, storage.Address, option)
+
+						labelMap, labelErr := sliceQuery.LabelMap()
+						if labelErr == nil {
```
这里的 labelErr 是不打算处理吗？

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+					})
+
+					if err != nil {
+						sendWg.Done()
```
为什么这里还需要再 done 一遍？

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+
+	total, list, resultTableOptions, done, err := queryRawWithScroll(s.ctx, queryTs)
+
+	log.Infof(s.ctx, "Test result - Total: %d, List length: %d, Done: %v, Error: %v", total, len(list), done, err)
```
log 同上。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+		}
+	}`
+
+	mock.Es.Set(map[string]any{
```
这个 kv 是有任何地断言的地方吗，key 任意写其他内容这里的单测会出错吗？

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+	assert.NotNil(s.T(), list)
+	assert.NotNil(s.T(), resultTableOptions)
+
+	if err != nil {
```
如果 err 是可以被预期的，建议断言会比 log 约束性更强。

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+		assert.Equal(s.T(), 0, len(resultTableOptions), "Expected empty resultTableOptions for non-existent table")
+	}
+
+	log.Infof(s.ctx, "Error handling test completed - Total: %d, List length: %d", total, len(list))
```
同上，没必要在每个测试跑完测试的时候都告诉世界我跑完了。

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+
+	// 验证 scroll session 管理正常
+	if len(list3) == 0 {
+		log.Infof(s.ctx, "[TEST] Scroll session completed successfully (empty final result)")
```
不要靠输出日志来验证你的测试是成功的，没有任何意义。

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+	}
+
+	log.Infof(s.ctx, "[TEST] Found %d unique gseIndex values", len(uniqueIds))
+	assert.GreaterOrEqual(s.T(), len(uniqueIds), 2, "Should have data from multiple slices")
```
为什么这里还可能是大于等于，是有可能是 2 也有可能是 3？

Zhaoyikaiii replied:
已移除.这部分测试用例已重写

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -330,7 +330,22 @@ func (i *Instance) esQuery(ctx context.Context, qo *queryOption, fact *FormatFac
 		if qb.Scroll != "" {
 			span.Set("query-scroll", qb.Scroll)
-			res, err = client.Scroll(qo.indexes...).Scroll(qb.Scroll).SearchSource(source).Do(ctx)
+			scroll := client.Scroll(qo.indexes...).Scroll(qb.Scroll).SearchSource(source)
```
<img width="935" height="436" alt="image" src="https://github.com/user-attachments/assets/c4bd0215-d649-4b0d-aafd-8a968f66688c" />

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -330,7 +330,22 @@ func (i *Instance) esQuery(ctx context.Context, qo *queryOption, fact *FormatFac
+				option := qb.ResultTableOptions.GetOption(qb.TableID, qo.conn.Address)
+				if option != nil {
+					if option.ScrollID != "" {
+						span.Set("query-scroll-id", option.ScrollID)
```
如果 scrollID 存在的情况下，scliceid slicemax 反而不用记录了？

chenjiandongx commented on pkg/unify-query/service/http/query_scroll_test.go:
```diff
@@ -0,0 +1,759 @@
+	log.Infof(s.ctx, "[TEST] All gseIndexes found: %v", allGseIndexes)
+
+	// 应该至少有一些数据
+	assert.Greater(s.T(), len(allGseIndexes), 0, "Should have collected some data")
```
所以数据量本身是随机的？

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -943,3 +944,301 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+		span.Set("query-highlight", queryTs.HighLight)
+
+		var hlF *function.HighLightFactory
+		if queryTs.HighLight != nil && queryTs.HighLight.Enable && len(allLabelMap) > 0 {
```
下载没有高亮的逻辑，不需要处理，直接对数据进行合并即可

Zhaoyikaiii replied:
已移除

shamcleren committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

Zhaoyikaiii committed:
Refactor scroll query handling: remove old test file and implement new scroll query logic

Zhaoyikaiii committed:
refactor: update ScrollSession to use structured SliceStatus for
ScrollIDs and SliceStatus maps

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,250 @@
+	return "", currentIndex, nil
+}
+
+func (s *ScrollSession) getNextElasticsearchScrollID(connect, tableID string, sliceIdx int) (string, int, error) {
```
看起来不存在返回 error 的情况。

Zhaoyikaiii replied:
已移除.

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+func (s *ScrollSession) SetScrollID(connect, tableID, scrollID string, sliceIdx int) {
+	if s.ScrollIDs == nil {
```
像这种直接假定已经初始即可，你已经给了一个 New 方法了。

Zhaoyikaiii replied:
已修改.已移除

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+func (s *ScrollSession) MarkSliceDone(connect, tableID string, sliceIdx int) {
+	if s.SliceStatus == nil {
```
同上，其他内容自检。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,178 @@
+	return err
+}
+
+func ScrollProcessSliceResult(ctx context.Context, sessionKey string, session *ScrollSession, connect, tableID, scrollID string, sliceIdx int, tsDbType string, size int64, options metadata.ResultTableOptions) error {
```
这个函数的入参数量是不是有点多？

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,178 @@
+	if err == nil {
+		return false
+	}
+	return err.Error() == "redis: nil"
```
这种 lib 本身是没有定义类似 ErrNil 这类的变量吗？如果后面升级依赖版本会不会就判断出错了。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func ScrollReleaseRedisLock(ctx context.Context, lock interface{}) error {
+	client := globalInstance.client
```
为什么这里就不需要判断 client 的合法性，`ScrollGetOrCreateSession` 就需要？

Zhaoyikaiii replied:
抱歉，已修改

chenjiandongx replied:
<img width="811" height="639" alt="image" src="https://github.com/user-attachments/assets/8b6458d7-a757-4fb3-811c-f5704beeb692" />

那这个看起来也没判断。

是否抽象一个函数用于获取全局的 Client。

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,178 @@
+		return err
+	}
+
+	err = client.Set(ctx, sessionKey, sessionBytes, session.ScrollTimeout).Err()
```
return 它即可。

Zhaoyikaiii replied:
已修改

chenjiandongx replied:
<img width="820" height="551" alt="image" src="https://github.com/user-attachments/assets/31b5e011-e274-4f63-95f5-7440e45e7f90" />

是有代码还没有推上来吗？

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,178 @@
+	return client.Set(ctx, key, sessionBytes, session.ScrollTimeout).Err()
+}
+
+func UpdateSession(ctx context.Context, sessionKey string, session ScrollSession) error {
```
这个函数与 `scrollUpdateSession` 函数是不是在做同一件事情。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,178 @@
+			}
+		}
+
+		if newScrollID != "" {
```
newScrollID 是否是在 isDone 为 false 的情况下才有意义，不然的话刚标记完 scrollid 后者就 remove。

如果是的话，是否可以优先处理 isDone 的逻辑。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -956,3 +957,186 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+
+	unit, start, end, timeErr := function.QueryTimestamp(queryTs.Start, queryTs.End)
+	if timeErr != nil {
+		return 0, nil, nil, false, timeErr
```
这种 error 返回方式还是要统一下，这代码读起来很别扭。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -956,3 +957,186 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+			ql.KeepColumns = queryTs.ResultColumns
+		}
+
+		qm, qmErr := ql.ToQueryMetric(ctx, queryTs.SpaceUid)
```
这种也是。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,353 @@
+	SessionLock *sync.Mutex
+}
+
+func NewSliceQueryContext(
```
一个函数 15 个入参？？

Zhaoyikaiii replied:
已修改,移除多余的ScrollQueryExecutor参数的传递

Zhaoyikaiii committed:
refactor: make test more clarify

Zhaoyikaiii committed:
refactor: renaming NewScrollQueryExecutor -> newScrollQueryExecutor

Zhaoyikaiii committed:
fix: missing err track

Zhaoyikaiii committed:
refactor: change error management

Zhaoyikaiii committed:
refactor: code structure

Zhaoyikaiii committed:
refactor: improve scroll session data handling and test clarity

Zhaoyikaiii committed:
chore: modify scroll test

Zhaoyikaiii committed:
fix(scroll): use deterministic MD5 key for scroll cache and set done flag

- Replace JSON-marshaled key with MD5 hash of sorted query data to ensure consistent cache keys
- Add missing `done = true` assignment when scroll is already finished
- Update scroll tests to match new scroll_id values

Zhaoyikaiii committed:
feat(json): add StableMarshal function for deterministic JSON serialization

Zhaoyikaiii committed:
refactor(redis): simplify return types in getNextElasticsearchScrollID and UpdateSession function

Zhaoyikaiii committed:
feat(scroll): enhance Elasticsearch slice handling and add Doris flow test

Zhaoyikaiii committed:
refactor(scroll): simplify slice creation and add spaceUid fallback

- Remove error returns from getNextDorisIndex, makeElasticsearchSlices and makeDorisSlices
- Add spaceUid fallback in queryRawWithScroll when empty
- Add jinzhu/copier and promote spf13/cast to direct dependency

Zhaoyikaiii committed:
fix(scroll): ensure session lock is properly released in processSliceQueryWithHelper

Zhaoyikaiii committed:
fix(redis): handle error return in UpdateSession function

Zhaoyikaiii committed:
refactor(metadata): remove JSON tags from SliceID and SliceMax fields in ResultTableOption

Zhaoyikaiii committed:
fix(metadata): add JSON tags to SliceID and SliceMax fields in ResultTableOption

Zhaoyikaiii committed:
fix: improve error handling and logging in scroll queries

- Return early on marshal error in queryRawWithScroll
- Add warning log when slice query submission fails

Zhaoyikaiii committed:
fix(scroll): handle scroll ID for ES and BkSql storage types

Only set scroll ID for Elasticsearch when non-empty; keep empty marker
for BkSql to indicate unfinished scroll.

Zhaoyikaiii committed:
chore: change err style

Zhaoyikaiii committed:
refactor: simplify return statements and reduce local vars in query/scroll

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,236 @@
+func (s *ScrollSession) HasMoreData(tsDbType string) bool {
+	switch tsDbType {
+	case consul.ElasticsearchStorageType:
+		if s.SliceStatus == nil {
```
Tips: 像这种判定也可以去除，写不写都不会影响逻辑。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,186 @@
+	if err == nil {
+		return false
+	}
+	return errors.Is(err, redis.Nil)
```
理论上这个分支会覆盖上一个分支，你可以试一下。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	result := client.Get(ctx, sessionKey)
+	if result.Err() == nil {
```
这里其实更优雅的方式应该是先判断是否为 Nil，剩下的就是正常的写法了。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func scrollUpdateSession(ctx context.Context, key string, session *ScrollSession) error {
+	client := globalInstance.client
```
像这种也是。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,339 @@
+	total              int64
+	message            strings.Builder
+
+	lock        *sync.Mutex
```
一般 lock 可以使用 struct{}，不需要指针，因为它是支持零值使用的。可以免去一些初始化步骤。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,339 @@
+		log.Warnf(queryCtx.Ctx, "Failed to process slice result: %v", err)
+		return err
+	}
+	result := SliceQueryResult{
```
这个 result 存在的目的是什么？

Zhaoyikaiii replied:
会通过`queryCtx.ResultCh`返回`queryCtx`下面每个`slice`的查询结果

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,339 @@
+	return nil
+}
+
+func (e *ScrollQueryExecutor) processQueryForStorage(
```
这个函数建议返回 err，err 内容由调用方统一处理。反正 error 都是要终止逻辑。

Zhaoyikaiii replied:
已添加

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,339 @@
+) {
+	connect := storage.Address
+	tableId := qry.TableID
+	instance := prometheus.GetTsDbInstance(e.ctx, qry)
```
为什么在下载日志的需求里，会有一个 `prometheus.GetTsDbInstance`？

Zhaoyikaiii replied:
这里是因为Unify-query操作下层的storage(es,doris,vm,prometheus...)都是通过这里的`prometheus.GetTsDbInstance`获取到`tsdb.Instance`这个接口的实现.
而我们现在实现的日志下载需要调用这里面的`QueryRawData`这个方法

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,339 @@
+	if options != nil {
+		resultTableOptions.MergeOptions(options)
+	}
+	*queryCtx.Total += size
```
这个 Total 为什么需要设计为 pointer？

Zhaoyikaiii replied:
已修改,进行了改造@Zhaoyikaiii
[refactor(scroll): replace shared state with result channel for concur…](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/853/commits/1577c02a94b2e5fb7d555918bc70e0559005c2b5)

chenjiandongx commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,339 @@
+	defer func() {
+		e.sendWg.Wait()
+		close(e.dataCh)
+		close(e.resultCh)
```
这个 resultch 看起来没有任何一个地方使用到。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,236 @@
+	SessionStatusFailed  = "FAILED"
+)
+
+func (s *ScrollSession) getNextDorisIndex() int {
```
getNextDorisIndex 返回的是 `currentIndex` 是符合预期的吗？

Zhaoyikaiii replied:
符合预期,原因是这里是后加加.其实语义上也是nextIndex

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,236 @@
+		TableID:  tableID,
+		SliceIdx: sliceIdx,
+	}
+	scrollID, exist := s.ScrollIDs[k.String()]
```
Tips: 像这种直接返回 `return s.ScrollIDs[k.String()]` 也是一样的。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/scroll_test.go:
```diff
@@ -0,0 +1,341 @@
+		assert.Equal(t, c.done, done, "Done should match expected value for step %d", i+1)
+		assert.Equal(t, c.hasData, hasData, "HasData should match expected value for step %d", i+1)
+
+		if t.Failed() {
```
这个可以不用，因为上面已经明确指出哪里可能会不符合预期了。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/service/http/scroll_test.go:
```diff
@@ -0,0 +1,341 @@
+		assert.Equal(t, c.done, done, "Done should match expected value for step %d", i+1)
+		assert.Equal(t, c.hasData, hasData, "HasData should match expected value for step %d", i+1)
+
+		if t.Failed() {
```
同上。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/service/http/scroll_test.go:
```diff
@@ -0,0 +1,341 @@
+	ir "github.com/TencentBlueKing/bkmonitor-datalink/pkg/utils/router/influxdb"
+)
+
+func TestQueryRawWithScroll_ESFlow(t *testing.T) {
```
这里看看需不需要补充一些【异常】的场景。

shamcleren replied:
+1

Zhaoyikaiii committed:
refactor(redis): remove redundant nil check for SliceStatus in HasMoreData

Zhaoyikaiii committed:
refactor(scroll): simplify Redis client usage and remove unused resultCh

- Replace local `client` vars with direct `globalInstance.client` access
- Remove unused `resultCh` field from ScrollQueryExecutor and sliceQueryContext
- Inline error handling in processQueryForStorage
- Clean up redundant nil checks for globalInstance

Zhaoyikaiii committed:
refactor: simplify scroll ID retrieval and inline call in makeElasticsearchSlices

Zhaoyikaiii committed:
refactor(scroll): replace shared state with result channel for concurrency safety

- Replace mutex-protected Total/Message fields with a dedicated resultCh
- Stream error handling via SliceQueryResult.Error/Message instead of strings.Builder
- Simplify ScrollQueryExecutor by removing lock and total/message state
- Ensure all slice-level errors are propagated through the new channel

Zhaoyikaiii committed:
refactor(redis): simplify ScrollGetOrCreateSession flow and remove redundant nil check

Zhaoyikaiii committed:
fix(scroll): prevent deadlock by closing result channel asynchronously

Zhaoyikaiii committed:
fix: wait for query execution to finish before collecting results

Zhaoyikaiii committed:
fix(scroll): prevent race condition when draining result channel

Move result collection into a goroutine and synchronize with a done
channel to ensure all results are processed before returning.

Zhaoyikaiii committed:
chore: add return empty data array

Zhaoyikaiii committed:
refactor(scroll): remove nil-client checks and unify error handling

- Drop redundant redis client nil checks in scroll_op.go
- Replace pkg/errors with stdlib errors in query.go
- Consolidate goroutine wait-groups and propagate errors via errors.Join
- Return explicit errors instead of named return vars in scroll.go

Zhaoyikaiii committed:
chore: add scroll trace

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,171 @@
+	if err := json.Unmarshal([]byte(result.Val()), &session); err != nil {
+		session = NewScrollSession(maxSlice, timeout, limit)
+	} else {
+		session.LastAccessAt = time.Now()
```
这个分支下面的函数会覆盖。

Zhaoyikaiii replied:
已移除

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,171 @@
+		return err
+	}
+
+	return globalInstance.client.Set(ctx, key, sessionBytes, session.ScrollTimeout).Err()
```
建议统一处理 globalClient 的获取逻辑。

Zhaoyikaiii replied:
修改为Client()

chenjiandongx commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,171 @@
+	return globalInstance.client.Set(ctx, key, sessionBytes, session.ScrollTimeout).Err()
+}
+
+func UpdateSession(ctx context.Context, sessionKey string, session *ScrollSession) (err error) {
```
建议当只有一个返回值时，不要用返回变量参数。

Zhaoyikaiii replied:
已修改

chenjiandongx commented on pkg/unify-query/service/http/query.go:
```diff
@@ -956,3 +957,205 @@ func QueryTsClusterMetrics(ctx context.Context, query *structured.QueryTs) (inte
+	return
+}
+
+func buildStorageQueryMap(queryList []*metadata.Query) (storageQueryMap map[string][]*metadata.Query) {
```
单个返回值建议在函数中进行初始化操作。

Zhaoyikaiii replied:
已修改

Zhaoyikaiii committed:
refactor(redis): unify error handling for lock acquisition and release

Zhaoyikaiii committed:
chore: add miss return

Zhaoyikaiii committed:
fix(scroll): send error result and update session on slice query failure

Ensure the scroll session is properly cleaned up and an error result is
emitted when a slice query fails, preventing stuck sessions and giving
callers visibility into failures.

Zhaoyikaiii committed:
Revert "fix(scroll): send error result and update session on slice query failure"

This reverts commit 11be68b03e1c80523352ab8eaab9f4f5a8d17914.

Zhaoyikaiii committed:
Revert "refactor(redis): unify error handling for lock acquisition and release"

This reverts commit 4451fd673bd6180e8cb846f14371b1ca475998e8.

Zhaoyikaiii committed:
Revert "chore: add miss return"

This reverts commit 56d5893376ff3a8f5b5accd5474a92e46dd2dd83.

Zhaoyikaiii committed:
Revert "chore: add scroll trace"

This reverts commit bc8f6336fcbd7c16787b813dae07866849c654e4.

Zhaoyikaiii committed:
Revert "chore: add return empty data array"

This reverts commit 14c3fd96c82d353cbd998c2ec77dd7bf3565a0f3.

Zhaoyikaiii committed:
Revert "refactor(scroll): remove nil-client checks and unify error handling"

This reverts commit 2d4ac19d753ba48291ee522b177a8a2968a20c2b.

Zhaoyikaiii committed:
Revert "fix(scroll): prevent race condition when draining result channel"

This reverts commit dc62b9e28163596528b0e6e1286cf1b431cc56e1.

Zhaoyikaiii committed:
Revert "fix: wait for query execution to finish before collecting results"

This reverts commit c1258016e3776a5850b0e3a86d6908d9aa55c75e.

Zhaoyikaiii committed:
Revert "refactor(redis): simplify ScrollGetOrCreateSession flow and remove redundant nil check"

This reverts commit ff61beefa6ada6261286f67c7d907217ae77d2f5.

Zhaoyikaiii committed:
Revert "fix(scroll): prevent deadlock by closing result channel asynchronously"

This reverts commit e4003f355015c323e2926036d2f22ddc5f75c5f0.

Zhaoyikaiii committed:
Revert "refactor: simplify scroll ID retrieval and inline call in makeElasticsearchSlices"

This reverts commit 77d553320eec4b899ed3822c816a1473ae36d81d.

Zhaoyikaiii committed:
Revert "refactor(scroll): replace shared state with result channel for concurrency safety"

This reverts commit 1577c02a94b2e5fb7d555918bc70e0559005c2b5.

Zhaoyikaiii committed:
Revert "refactor(redis): remove redundant nil check for SliceStatus in HasMoreData"

This reverts commit bdd9e856b32bbcdb13917e0649e9a6956ff923ab.

Zhaoyikaiii committed:
Revert "refactor(scroll): simplify Redis client usage and remove unused resultCh"

This reverts commit b7015b974e7300e9750faef4776efdc691937e62.

Zhaoyikaiii committed:
Revert "refactor: simplify return statements and reduce local vars in query/scroll"

This reverts commit b988c2ed37ec9bb05d95e26dd06622ea6b557953.

Zhaoyikaiii committed:
Revert "chore: change err style"

This reverts commit 809ff1a5571cf44e572e1213202594417c331c72.

Zhaoyikaiii committed:
Revert "fix: improve error handling and logging in scroll queries"

This reverts commit 90a15594e4e43862a1d86721b2d610881a437663.

Zhaoyikaiii committed:
Revert "fix(scroll): handle scroll ID for ES and BkSql storage types"

This reverts commit 32941aaf11f134291f87fedcddc52d6417e51c77.

Zhaoyikaiii committed:
Revert "fix(metadata): add JSON tags to SliceID and SliceMax fields in ResultTableOption"

This reverts commit c598f054c44c1f23cd18d26fde328fabfe61afde.

Zhaoyikaiii committed:
Revert "refactor(metadata): remove JSON tags from SliceID and SliceMax fields in ResultTableOption"

This reverts commit 274a8783b1468cb95bbb5c4d2b8490a8c99879f6.

Zhaoyikaiii committed:
refactor(scroll): replace slice status map with set-based scroll ID tracking

- Replace ScrollIDs map[string]string and SliceStatus map[string]bool with
  map[string]map[string]struct{} to store multiple scroll IDs per slice
- Add github.com/deckarep/golang-set dependency for set operations
- Update scroll session logic to handle multiple scroll IDs per key
- Refactor Redis lock functions to use Client() helper consistently
- Adjust tests to reflect new scroll ID structure

Zhaoyikaiii committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

Zhaoyikaiii committed:
chore: done return empty list data

Zhaoyikaiii committed:
chore: add esQuery retry

Zhaoyikaiii committed:
fix: scroll提前结束

Zhaoyikaiii committed:
feat(scroll): add retry mechanism for failed slices and refactor headers

- Introduce slice-level retry logic with configurable max attempts
- Add scroll_retry_test.go for retry scenario coverage
- Refactor metadata.Headers to merge instead of overwrite
- Remove unused deps: deckarep/golang-set, spf13/cast (indirect)
- Reorder json.Marshal/StableMarshal for clarity
- Update ES & BK-SQL drivers to support new scroll retry states

shamcleren commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,67 @@
+	"time"
+)
+
+func AcquireScrollSessionLock(ctx context.Context, sessionKeySuffix string, dur time.Duration) error {
```
可以把这个实现抽象一层，放到 redis.go 文件下

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/service/http/handler.go:
```diff
@@ -251,6 +252,95 @@ func HandlerQueryRaw(c *gin.Context) {
+	}
+	defer func() {
+		if err = redis.ReleaseScrollSessionLock(ctx, sessionLockKey); err != nil {
+			log.Errorf(ctx, "release scroll session lock error: %s", err.Error())
```
这里可以直接去掉

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -413,6 +359,76 @@ func queryRawWithInstance(ctx context.Context, queryTs *structured.QueryTs) (tot
+	go func() {
+		defer readerWg.Done()
+		for d := range dataCh {
+			listMutex.Lock()
```
这里的锁应该不需要

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -413,6 +359,76 @@ func queryRawWithInstance(ctx context.Context, queryTs *structured.QueryTs) (tot
+				sData, sErr := scrollQueryWorker(ctx, session, e.Connect, e.TableID, qry, start, end, e.Storage.Instance)
+				if sErr != nil {
+					messageMutex.Lock()
+					message.WriteString(fmt.Sprintf("query %s:%s is error: %s ", qry.TableID, qry.Fields, sErr.Error()))
```
这里的错误可以使用 channel 的方式实现

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -413,6 +359,76 @@ func queryRawWithInstance(ctx context.Context, queryTs *structured.QueryTs) (tot
+	readerWg.Wait()
+
+	total = int64(len(list))
+	done = session.CouldDone()
```
命名可以直接改成 Done()

Zhaoyikaiii replied:
已修改

Zhaoyikaiii committed:
refactor(scroll): unify lock helpers and clean up scroll session code

- Replace Acquire/ReleaseScrollSessionLock with generic AcquireLock/ReleaseLock
- Move lock key prefix out of scroll package into redis package
- Rename CouldDone to Done for clarity
- Remove unused CalcStorageIDs method from Query struct
- Clean up scroll key generation and naming

shamcleren commented on pkg/unify-query/redis/redis.go:
```diff
@@ -47,6 +48,16 @@ func Client() goRedis.UniversalClient {
 }
 
+func IsNil(err error) bool {
+	if err == nil {
```
这里的判断和函数命名含义有冲突吧？看上去是需要判断 IsRedisNil ？

Zhaoyikaiii replied:
哦,因为这里是redis包下面.外面调用的时候一般都是redis.IsNil 这样.

Zhaoyikaiii replied:
可能IsNil也有问题,可能表示为redis.IsEntryNil这样可能会比IsNil好一些

shamcleren commented on pkg/unify-query/service/http/scroll_retry_test.go:
```diff
@@ -0,0 +1,217 @@
+	return ""
+}
+
+func generateScrollSliceStatusKey(storageType, connect, tableID string, sliceIdx int) string {
```
这里可以更一下入参，方便扩展，例如：

generateScrollSliceStatusKey(item ...string) string，如果要支持多种类型的话，可以用 ...any，然后结合 cast.ToString()

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,272 @@
+	Offset      int
+}
+
+func (s *ScrollSession) MakeSlices(storageType, connect, tableID string) (slices []*SliceInfo, err error) {
```
这个问题应该说了好几次了，返回的方式不要两种混合使用，要么就不用声明，直接返回内容，要么就提前声明返回空

Zhaoyikaiii replied:
已修改,移除了返回值列表中的变量声明

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,272 @@
+		key := generateScrollSliceStatusKey(consul.ElasticsearchStorageType, connect, tableID, i)
+		val, exists := s.ScrollIDs[key]
+
+		if !exists {
```
这里的位置可以调整一下，终止放前面（失败超过次数/结束），可执行放后面，另外，这里更新 redis 是不是没有必要，最终执行完毕之后，才更新吧？备注也需要写清楚一点，这里对状态值的判定，应该只需要判定是否需要执行就行

shamcleren replied:
这里需要继续修改一下，按照状态处理，最终统一更新

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,272 @@
+		})
+	}
+
+	if needUpdate {
```
这个更新 es 和 doris 应该在一起

shamcleren commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,293 @@
+		isEmptyData := len(data) == 0 && total == 0
+		isScrollIDEmpty := sliceResultOption != nil && sliceResultOption.ScrollID == ""
+
+		if isEmptyData && oldSliceResultOption != nil && oldSliceResultOption.SliceIndex != nil {
```
从这里到下面，这里的结束回写状态过于复杂了，到这一层级应该就是三种状态：成功，失败，结束，UpdateScrollID 这个函数应该只要执行一遍

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,272 @@
+	return fmt.Sprintf("%s:%s:%s:%d", storageType, connect, tableID, sliceIdx)
+}
+
+func NewScrollSession(sessionKeySuffix string, maxSlice int, scrollTimeout time.Duration, limit int) *ScrollSession {
```
这里是否可以用接口把 es 和 doris 抽象开，而不是粗暴的用不同的函数调用？后续如果下载接口需要适配其他类型，也容易做

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/redis/scroll_op.go:
```diff
@@ -0,0 +1,64 @@
+	return session, nil
+}
+
+func checkScrollSession(ctx context.Context, key string) (session *ScrollSession, exist bool, err error) {
```
这里都是一样的问题，两套 return 方案用在同一个函数里面

Zhaoyikaiii replied:
已修改,移除了返回值列表中的变量初始化

shamcleren commented on pkg/unify-query/service/http/handler.go:
```diff
@@ -251,6 +252,97 @@ func HandlerQueryRaw(c *gin.Context) {
+	}
+
+	queryStr, _ := json.StableMarshal(queryTs)
+	scrollKey, err := generateScrollKey(user.Name, *queryTs)
```
这里上面已经有一个 queryStr 了，是不是可以直接复用？

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -406,10 +365,96 @@ func queryRawWithInstance(ctx context.Context, queryTs *structured.QueryTs) (tot
+				}
+			}
+
+			if dataCh == nil && errCh == nil {
```
其实这里可以单独放一个 recv 作为结束信号

shamcleren commented on pkg/unify-query/tsdb/bksql/instance.go:
```diff
@@ -240,8 +240,16 @@ func (i *Instance) QueryRawData(ctx context.Context, query *metadata.Query, star
 		option = &metadata.ResultTableOption{}
 	}
+
+	resultOption := &metadata.ResultTableOption{
```
这里传过来的意义是啥？为啥要改？

Zhaoyikaiii replied:
这个相当于在QueryRawData内部进行了一次拷贝操作.因为这里的scroll是并发的,会存在同时操作一个resultOption的情况,所以在内部单独进行了一次拷贝

shamcleren replied:
这里需要区分 queryOption 和 resultOption

Zhaoyikaiii committed:
refactor(scroll): improve scroll key generation and function signatures

- Refactor generateScrollSliceStatusKey function to support variadic parameters and better nil handling
- Unify function return signature style by removing named return values
- Use cast.ToString for type conversion to avoid fmt.Sprintf performance overhead
- Optimize code structure for better readability and maintainability

Zhaoyikaiii committed:
feat(scroll): implement ScrollHandler interface and update slice handling logic

Zhaoyikaiii committed:
fix(scroll): add mutex and atomic updates to prevent race conditions

Introduce sync.RWMutex in ScrollSession to protect concurrent access to
ScrollIDs map. Replace direct map updates with new AtomicUpdateSliceStatus
and AtomicUpdateSliceStatusAndOffset helpers that lock, update, and
persist atomically. Update scroll query worker to use these helpers and
improve option retrieval logic for different storage types.

Zhaoyikaiii committed:
refactor(scroll): rename loop variable and update slice offset calculation

Zhaoyikaiii committed:
refactor(scroll): simplify scroll status update logic and enhance key generation

Zhaoyikaiii committed:
refactor(scroll): recovery option handling and remove redundant variable assignments

Zhaoyikaiii committed:
refactor(scroll): update session creation logic to return early on error

Zhaoyikaiii committed:
refactor(scroll): log query body in span for better traceability

Zhaoyikaiii committed:
refactor(scroll): add instance assertion for DefaultInstance

shamcleren commented on pkg/unify-query/service/http/scroll.go:
```diff
@@ -0,0 +1,293 @@
+	return
+}
+
+func collectStorageQuery(qryList []*metadata.Query) map[string][]*metadata.Query {
```
这个函数实现有问题，不是所有查询都有 storageID， 需要参考 GetTsDbInstance 函数实现

Zhaoyikaiii committed:
refactor(scroll): enhance instance connection handling and simplify scroll status logic

Zhaoyikaiii committed:
refactor(scroll): remove unused CollectStorageIDs method from Query struct

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,151 @@
+	return true
+}
+
+func (s *ScrollSession) UpdateSliceStatus(ctx context.Context, sliceKey string, status string, scrollID string) error {
```
这个函数跟下面那个函数可以合并成一个函数

Zhaoyikaiii replied:
已修改

Zhaoyikaiii committed:
refactor(scroll): consolidate slice status update logic and remove redundant method

Zhaoyikaiii committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

Zhaoyikaiii committed:
feat: enhance scroll session management and improve query handling

Zhaoyikaiii committed:
chore: 更新scroll状态

Zhaoyikaiii committed:
chore: 修改日志下载

Zhaoyikaiii committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

Zhaoyikaiii committed:
feat: refactor handler registration and restore missing configuration path

shamcleren commented on pkg/unify-query/internal/json/json.go:
```diff
@@ -26,6 +26,17 @@ var sonicAPI = sonic.Config{
 	CopyString:       true, // 正确性
 }.Froze()
 
+var stableSonicAPI = sonic.Config{
```
这个在原函数基础上改，我已经改了，并且合并到主干分支了，你拉下 master 合并下就可以看到

Zhaoyikaiii replied:
已移除

Zhaoyikaiii committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,185 @@
+)
+
+type ScrollSession struct {
+	Ctx               context.Context    `json:"-"`
```
这个 Ctx 好像没什么作用？Redis Client 调用的那个 Ctx 完全可以使用函数的参数来传，没必要放到这里

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+func checkScrollSession(ctx context.Context, queryTsStr string) (*ScrollSession, bool) {
+	var session ScrollSession
```
优化一下写法，提前定义就行了

func checkScrollSession(ctx context.Context, queryTsStr string) (*ScrollSession, bool) {
	session := &ScrollSession{}
	err := Client().Get(ctx, SessionKeyPrefix+queryTsStr).Scan(session)
	if err != nil {
		return nil, false
	}
	
	return session, true
}

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/service/http/handler.go:
```diff
@@ -251,6 +252,102 @@ func HandlerQueryRaw(c *gin.Context) {
+	span.Set("query-body", queryStr)
+
+	if isClearCache {
+		session.ReleaseLock()
```
这个函数直接支持 ctx 就不需要在 session 的结构体中定义 ctx 了

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -406,20 +359,146 @@ func queryRawWithInstance(ctx context.Context, queryTs *structured.QueryTs) (tot
+
+	queryRef.Range("", func(qry *metadata.Query) {
+		for k, s := range session.ScrollIDs {
+			if s.Status == redisUtil.StatusCompleted || s.FailedNum > session.SliceMaxFailedNum {
```
这个可以再往下实现一层，然后复用就行，例如：SliceStatusValue 实现一个 Done 函数，然后上层 ScrollSession 也能直接复用

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/service/http/query.go:
```diff
@@ -406,20 +359,146 @@ func queryRawWithInstance(ctx context.Context, queryTs *structured.QueryTs) (tot
+					errCh <- err
+					return
+				}
+				newQry.SliceID = s.ScrollID
```
这个可以一起放到 ResulTableOption 里面去定义了，外面 query 就可以去掉了，其他用到的地方也可以改成 ResultTableOption.SliceID，问题不大

shamcleren commented on pkg/unify-query/tsdb/bksql/instance.go:
```diff
@@ -79,15 +77,20 @@ func NewInstance(ctx context.Context, opt *Options) (*Instance, error) {
 	return instance, nil
 }
 
+func (i *Instance) Connect() string {
```
这个函数应该没在用了吧？可以不需要了

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/tsdb/bksql/instance.go:
```diff
@@ -509,6 +509,10 @@ func (i *Instance) InstanceType() string {
 	return consul.BkSqlStorageType
 }
 
+func (i *Instance) IsCompleted(opt *metadata.ResultTableOption, dataLen int) bool {
```
这个也没看到哪里有调用啊？

Zhaoyikaiii replied:
已移除

shamcleren commented on pkg/unify-query/tsdb/bksql/client.go:
```diff
@@ -45,6 +45,10 @@ func (c *Client) WithHeader(headers map[string]string) *Client {
 	return c
 }
 
+func (c *Client) URL() string {
```
这里也没东西调用

Zhaoyikaiii replied:
已移除

Zhaoyikaiii committed:
refactor: Remove unused methods and fields, update method signatures

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,186 @@
+func (s *ScrollSession) ReleaseLock(ctx context.Context) {
+	s.Mu.Lock()
+	defer s.Mu.Unlock()
+	Client().Del(ctx, s.LockKey).Err()
```
Del / Set 操作如果有 error 是不需要处理吗

Zhaoyikaiii replied:
已修改

Zhaoyikaiii committed:
fix: Improve lock release error handling

chenjiandongx commented on pkg/unify-query/query/structured/query_ts.go:
```diff
@@ -613,45 +629,105 @@ func (q *Query) ToQueryMetric(ctx context.Context, spaceUid string) (*metadata.Q
-					return true
+		for _, storageID := range storageIDs {
+			query, buildErr := q.BuildMetadataQuery(ctx, tsDB, allConditions)
+			if buildErr != nil {
```
这个 err 也处理下

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,193 @@
+	Limit             int                `json:"limit"`
+	ScrollIDs         []SliceStatusValue `json:"scroll_ids"`
+
+	Mu sync.RWMutex `json:"-"`
```
这个 Mu 应该是小写会好一些，内部变量，不宜对外暴露。

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,193 @@
+	return Client().Set(ctx, s.SessionKey, s, s.ScrollTimeout).Err()
+}
+
+func (s *ScrollSession) ReleaseLock(ctx context.Context) (err error) {
```
跟上面的函数保持统一风格吧。

chenjiandongx commented on pkg/unify-query/redis/scroll.go:
```diff
@@ -0,0 +1,193 @@
+	s.Mu.RLock()
+	defer s.Mu.RUnlock()
+	for _, sliceValue := range s.ScrollIDs {
+		if sliceValue.Status != StatusCompleted && sliceValue.FailedNum < s.SliceMaxFailedNum {
```
这个跟上面的 Done 的关系是什么，为什么上面的是个 DefaultSliceMaxFailedNum 常量。

Zhaoyikaiii replied:
- 这两个Done分别用来标记scrollSession和scrollSlice是否结束.
- 上面是个DefaultSliceMaxFailedNum 是因为他获取不到scrollSession的maxFailedNum这个值.
- 又因为scrollSlice也是从这个常量传递进来的,所以在上面直接使用了这个常量.
- 我可以修改为在scrollSlice里面也增加这个参数
- 这样可能更好一些,因为这个常量现在只有单一的变量引用

Zhaoyikaiii committed:
refactor: Clean up code and simplify logic

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -728,134 +684,54 @@ func (i *Instance) QuerySeriesSet(
-
-	go func() {
-		defer func() {
-			// es 查询有很多结构体无法判断的，会导致 panic
```
recover 逻辑被删除了？

Zhaoyikaiii replied:
没有被删除.现在在这个函数的开始位置.

Zhaoyikaiii committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

Zhaoyikaiii committed:
refactor: Update scroll session parameters for improved performance

shamcleren committed:
Merge branch 'master' into feat/scroll_query/#1010158081125200073

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073

shamcleren committed:
Merge remote-tracking branch 'Zhaoyikaiii/feat/scroll_query/#1010158081125200073' into feat/scroll_query/#1010158081125200073

# Conflicts:
#	pkg/unify-query/redis/scroll.go
#	pkg/unify-query/service/http/query_test.go

shamcleren committed:
feat: unify-query支持日志下载 #1010158081125200073
